### PR TITLE
Bug 1239722 - Open a new tab when new version is detected

### PR DESCRIPTION
**Pull request by bkmunar**
1. created function in AppConstants to return the newest version of the "What's New" page for iOS
2. imported Readability file into 3rd party folder (https://github.com/Isuru-Nanayakkara/Reach)
3. updated BVC viewDidAppear logic and added a network connectivity check before opening the new tab upon app update

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -35,7 +35,13 @@ private struct BrowserViewControllerUX {
     private static let BookmarkStarAnimationOffset: CGFloat = 80
 }
 
+protocol BrowserViewControllerDelegate: class {
+    func browserViewControllerAddReleaseNotesOnUpdate(browserViewController: BrowserViewController, profile: Profile)
+}
+
 class BrowserViewController: UIViewController {
+    var delegate: BrowserViewControllerDelegate?
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: `weak`

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -35,7 +35,13 @@ private struct BrowserViewControllerUX {
     private static let BookmarkStarAnimationOffset: CGFloat = 80
 }
 
+protocol BrowserViewControllerDelegate: class {
+    func browserViewControllerAddReleaseNotesOnUpdate(browserViewController: BrowserViewController, profile: Profile)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: The delegate shouldn't be aware of what clients are doing with the implementation, so you should use a generic event name (e.g., `browserViewControllerDidAppear`).

File: Client/Application/AppDelegate.swift [hunk 3]
```diff
@@ -441,6 +442,15 @@ extension AppDelegate: MFMailComposeViewControllerDelegate {
...
 }
 
+extension AppDelegate: BrowserViewControllerDelegate {
+    func browserViewControllerAddReleaseNotesOnUpdate(browserViewController: BrowserViewController, profile: Profile) {
+        if profile.prefs.stringForKey("latestAppVersion") != appVersion {
+            browserViewController.openURLInNewTab(NSURL(string: "https://www.mozilla.org/en-US/firefox/ios/\(appVersion)/releasenotes/")!)
+        }
+        profile.prefs.setString(appVersion, forKey: "latestAppVersion")
+    }
+}
```
Comment (thebnich) [inline Client/Application/AppDelegate.swift]: This isn't great -- we're making assumptions about the state of `BVC` (a view) to decide when to add a new tab (our model). Ideally, we'd tease apart the coupling of adding a tab and inserting the web view. This might be OK for now until we can do our BVC refactor.

**Comment by rnewman** [comment]
Move the app version accessor into `AppInfo.swift`. Then move this logic into `BrowserViewController`: one single block inside the view callback. If this is a hack, and we're going to have to futz with it later, then we might as well avoid spreading hack-juice over two separate files, defining delegate protocols etc. etc..

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -521,6 +527,8 @@ class BrowserViewController: UIViewController {
         log.debug("BVC calling super.viewDidAppear.")
         super.viewDidAppear(animated)
         log.debug("BVC done.")
+
+        delegate?.browserViewControllerDidAppear(self, profile: profile)
```
Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: So right here, just do:

```
let appVersion = AppInfo.appVersion
if profile.prefs.stringForKey("latestAppVersion") != appVersion {
    self.openURLInNewTab(AppConstants.whatsNewURLForVersion(appVersion))
    profile.prefs.setString(appVersion, forKey: "latestAppVersion")
}
```

Note that you only incur the prefs write if the version changed!

Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: Follow-up: check for network connectivity before trying to open the tab. You can use the various `Reachability` libraries to do this, or import `SystemConfiguration` yourself, or fetch the page contents here, or add a variant of `openURLInNewTab` that waits for a successful fetch before opening the tab and running a provided callback.

File: Client.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -4079,6 +4082,20 @@
 			remoteRef = 39A362D71AAF5E2C00F47390 /* PBXContainerItemProxy */;
 			sourceTree = BUILT_PRODUCTS_DIR;
 		};
+		74108D631C52BCF90076D651 /* WebImage.framework */ = {
```
Comment (rnewman) [inline Client.xcodeproj/project.pbxproj]: Get rid of all of these unrelated project file changes.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -521,6 +521,24 @@ class BrowserViewController: UIViewController {
         log.debug("BVC calling super.viewDidAppear.")
         super.viewDidAppear(animated)
         log.debug("BVC done.")
+
+        let appVersion = AppInfo.appVersion
+        if let whatsNewURL = AppConstants.whatsNewURLForVersion(appVersion) where profile.prefs.stringForKey("latestAppVersion") != appVersion && hasConnectivity() {
```
Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: Try running this in a Playground:

```
func x() -> Int? {
    print("I was called")
    return 5
}

func y() -> Bool {
    print("So was I")
    return false
}

if let xx = x() where y() {
    print("Here")
}
```

to see what's up with this, then change it to

```
if profile.prefs.stringForKey(…) != appVersion && hasConnectivity() {
    self.openURLInNewTab(AppConstants.whatsN…(…))
}
```

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 7]
```diff
@@ -521,6 +521,24 @@ class BrowserViewController: UIViewController {
...
+            self.openURLInNewTab(whatsNewURL)
+            profile.prefs.setString(appVersion, forKey: "latestAppVersion")
+        }
+    }
+
+    func hasConnectivity() -> Bool {
```
Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: Move this to `DeviceUtils`.

File: Utils/AppConstants.swift [hunk 8]
```diff
@@ -58,4 +58,12 @@ public struct AppConstants {
     return true
 #endif
     }()
+
+    /// Function returning the What's New page for iOS
+    public static func whatsNewURLForVersion(version: String) -> NSURL? {
+        if let whatsNewURL = NSURL(string: "https://support.mozilla.org/1/mobile/\(version)/%OS%/%LOCALE%/new-ios") {
+            return whatsNewURL
```
Comment (rnewman) [inline Utils/AppConstants.swift]: Just

```
return NSURL(string: "…")
```

— you're testing for an optional and returning the value or `nil`, which is redundant.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 6 above)
Comment (bkmunar) [inline Client/Frontend/Browser/BrowserViewController.swift]: omg rofl oops idk what my brain did there @__@

**Comment by bkmunar** [comment]
updated! @rnewman

File: Client.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -1020,13 +1022,6 @@
 			remoteGlobalIDString = EABD46AA1857A0C700A5F081;
 			remoteInfo = KIF;
 		};
-		D38B2D6D1A8D976A0040E6B5 /* PBXContainerItemProxy */ = {
```
Comment (rnewman) [inline Client.xcodeproj/project.pbxproj]: Kill this change…

File: Client.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -1041,13 +1036,6 @@
 			remoteGlobalIDString = EABD46CD1857A0F300A5F081;
 			remoteInfo = "KIF Tests";
 		};
-		D38B2D731A8D976A0040E6B5 /* PBXContainerItemProxy */ = {
```
Comment (rnewman) [inline Client.xcodeproj/project.pbxproj]: … and this one.

File: Client.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -2619,10 +2609,8 @@
 			isa = PBXGroup;
 			children = (
 				D38B2D6C1A8D976A0040E6B5 /* libKIF.a */,
-				D38B2D6E1A8D976A0040E6B5 /* libKIF-OCUnit.a */,
```
Comment (rnewman) [inline Client.xcodeproj/project.pbxproj]: And these, and the rest that aren't about Reachability.

File: Reachability.swift [hunk 12]
```diff
@@ -0,0 +1,113 @@
+// The MIT License (MIT)
```
Comment (rnewman) [inline Reachability.swift]: Put this file in `ThirdParty`, both the folder and the project group.

File: Utils/DeviceInfo.swift [hunk 13]
```diff
@@ -3,6 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import UIKit
+import Shared
```
Comment (rnewman) [inline Utils/DeviceInfo.swift]: You don't need this import: `DeviceInfo` is *in* `Shared`.

**Comment by bkmunar** [comment]
@rnewman lemme know how it looks!

File: Client.xcodeproj/project.pbxproj [hunk 14]
```diff
@@ -2131,7 +2134,7 @@
 				288A2DB41AB8B38D0023ABC3 /* Result.swift */,
 			);
 			name = Result;
-			sourceTree = "<group>";
+			sourceTree = SOURCE_ROOT;
```
Comment (rnewman) [inline Client.xcodeproj/project.pbxproj]: That doesn't seem right.

**Comment by rnewman** [comment]
Looks good apart from that potentially spurious project file change. Rebase onto current master and fix that, and we're good to go.

**Comment by bkmunar** [comment]
thanks!!
